overlord/snapshotstate/snapshotmgr.go

```diff
@@ -0,0 +1,277 @@
+	backendSave          = backend.Save
+	backendRestore       = (*backend.Reader).Restore // TODO: look into using an interface instead
+	backendCheck         = (*backend.Reader).Check
+	backendRevert        = (*backend.RestoreState).Revert // ditto
+	backendCleanup       = (*backend.RestoreState).Cleanup
+)
+
+// SnapshotManager takes snapshots of active snaps
+type SnapshotManager struct {
+	runner *state.TaskRunner
+}
+
+// Manager returns a new SnapshotManager
+func Manager(st *state.State) *SnapshotManager {
+	runner := state.NewTaskRunner(st)
+	runner.AddHandler("save-snapshot", doSave, doForget)
+	runner.AddHandler("forget-snapshot", doForget, nil)
+	runner.AddHandler("check-snapshot", doCheck, nil)
+	runner.AddHandler("restore-snapshot", doRestore, undoRestore)
+	runner.AddCleanup("restore-snapshot", cleanupRestore)
+	return &SnapshotManager{runner: runner}
+}
+
+// KnownTaskKinds returns a unsorted list of known task kinds.
+// It is part of the overlord.StateManager interface.
+func (m *SnapshotManager) KnownTaskKinds() []string {
+	return m.runner.KnownTaskKinds()
+}
+
+// Ensure is part of the overlord.StateManager interface.
+func (m *SnapshotManager) Ensure() error {
+	m.runner.Ensure()
+	return nil
+}
+
+// Wait is part of the overlord.StateManager interface.
+func (m *SnapshotManager) Wait() {
+	m.runner.Wait()
+}
+
+// Stop is part of the overlord.StateManager interface.
+func (m *SnapshotManager) Stop() {
+	m.runner.Stop()
+}
+
+type snapshotState struct {
+	SetID    uint64   `json:"set-id"`
+	Snap     string   `json:"snap"`
+	Users    []string `json:"users,omitempty"`
+	Filename string   `json:"filename,omitempty"`
+}
+
+func filename(setID uint64, si *snap.Info) string {
+	skel := &client.Snapshot{
+		SetID:    setID,
+		Snap:     si.Name(),
+		Revision: si.Revision,
+		Version:  si.Version,
+	}
+	return backend.Filename(skel)
+}
+
+// prepareSave does all the steps of doSave that require the state lock;
+// it has no real significance beyond making the lock handling simpler
+func prepareSave(task *state.Task) (snapshot *snapshotState, cur *snap.Info, cfg map[string]interface{}, err error) {
+	st := task.State()
+	st.Lock()
+	defer st.Unlock()
+
+	if err := task.Get("snapshot", &snapshot); err != nil {
+		return nil, nil, nil, err
+	}
+	cur, err = snapstateCurrentInfo(st, snapshot.Snap)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	snapshot.Filename = filename(snapshot.SetID, cur)
+	task.Set("snapshot", &snapshot)
+
+	rawCfg, err := configGetSnapConfig(st, snapshot.Snap)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	if rawCfg != nil {
+		if err := json.Unmarshal(*rawCfg, &cfg); err != nil {
+			return nil, nil, nil, err
+		}
+	}
+
+	return snapshot, cur, cfg, nil
+}
+
+func doSave(task *state.Task, tomb *tomb.Tomb) error {
+	snapshot, cur, cfg, err := prepareSave(task)
+	if err != nil {
+		return err
+	}
+	_, err = backendSave(tomb.Context(nil), snapshot.SetID, cur, cfg, snapshot.Users)
```

Comment:
I think more consistent would be to pass context.TODO() which we use universally everywhere (and serves us as an easy to spot reminder).

Reply:
I think using the tomb for context is the right thing to do when we have a tomb (and it's what we do in `snapstate` for the download context).

I'll take a look at getting rid of a lot of those TODOs in a separate PR though. The store seems to abusing them quite a bit.